### PR TITLE
Add Microsoft Lens (Turbo): GPT-OSS multi-layer encoder and 48-block MMDiT in MLX

### LINKED ISSUE #424: Feature Request: model support for Microsoft Lens
Microsoft has recently released a new 3.8B image generation model, "Lens" that uses GPT-OSS encoder and Flux2 VAE.

Early reviews look very promising, seemingly punching above its weight and a good choice for Macs with modest memory specs. 
 
https://github.com/microsoft/Lens
https://huggingface.co/microsoft/Lens-Base
https://huggingface.co/microsoft/Lens
https://huggingface.co/microsoft/Lens-Turbo

### PR DESCRIPTION
**Pull request by fxd0h**
Closes [#424](https://github.com/filipstrand/mflux/issues/424). Microsoft Lens (Turbo) running natively in MLX: the 3.8B dual-stream MMDiT with GPT-OSS 20B multi-layer text features and the FLUX.2 VAE already in-tree.

```bash
mflux-generate-lens --prompt "A cozy bookshop cafe at dusk, warm light through the window, a cat sleeping on a stack of books" --seed 7 --width 512 --height 512 --steps 4
```

lens sample, seed 7, 4 steps

How it is put together:

- Encoder: GPT-OSS 20B vendored from mlx-lm (MIT, five modules under `common_models/gpt_oss/`, sharding and KV-cache paths stripped, no new dependency). The Lens wrapper renders the harmony template with the reference's frozen date, captures the residual stream after blocks (5, 11, 17, 23) with no final norm, and drops the first 97 tokens. That offset is exactly the tokenized system+developer blocks plus the user message header, verified token by token, so the template is byte-compatible with the reference pipeline. Over-long prompts truncate the prompt segment only, keeping the frozen prefix and suffix aligned, and harmony control markers are stripped from user prompts.
- Denoiser: 48-block `LensTransformer` reusing `QwenEmbedRopeMLX` verbatim (identical centered-halves rope semantics, axes 8/28/28, confirmed against the reference at 2e-6) and the Qwen timestep embedding. The joint attention sequence puts the image stream first and `norm_out` is scale-first AdaLayerNormContinuous, both per the reference. Weight paths are 1:1 with the published checkpoint: 1264/1264 tensor coverage with no mapping layer.
- Family: `lens-turbo` / `lens` config, components resolving from their three homes (the Microsoft repositories were withdrawn, so the DiT comes from the Comfy-Org mirror, the encoder from the mlx-community mlx-format conversion, and the VAE from the klein repository through the existing flux2 loader). Flow-match sigmas use the reference scheduler's dynamic exponential shift. The `IGNORED_OPTIONS` dict on the CLI documents that `--guidance`/`--negative-prompt` are never applied (turbo is CFG-internalized); [#499](https://github.com/filipstrand/mflux/pull/499) would wire it into runtime warnings.

Numerical validation against the ComfyUI reference implementation, same injected inputs on both sides: rope identical to 2e-6, per-block cosine 0.99998, and end to end in float32 the two implementations agree at cosine 0.999658 while the reference's own bfloat16 run only reaches 0.9637 against its own float32. This port is closer to the fp32 ground truth than the reference is to itself at working precision.

Tests: 12 fast (frozen template strings, capture semantics vs a manual forward, the 97-token prefix with the real tokenizer, overflow truncation, marker stripping, weight-path conventions, config aliases) plus one slow golden. Heads up on the golden: the reference image was generated on an M5 Max and may need regeneration on your reference hardware. Full upstream fast suite passes at 525.

Scope: turbo checkpoint only for now (base Lens shares the exact tensor layout, so it is a weight swap); no LoRA, img2img, or stepwise previews yet.

## Summary by CodeRabbit

* **New Features**
  * Added Lens Turbo text-to-image generation with prompt encoding, seeded generation, configurable dimensions, inference steps, quantization, and metadata.
  * Added the `mflux-generate-lens` command-line tool.
  * Added Lens Turbo model configuration and GPT-OSS-based prompt support.
  * Added Lens transformer and mixture-of-experts model support, including rotary embeddings and optimized attention.

* **Tests**
  * Added CLI, encoder, transformer, and image-generation coverage, including reference-image regression tests.

### CODE REVIEW

**Comment by fxd0h** [comment]
@coderabbitai full review

**Comment by fxd0h** [comment]
@greptileai review

**Comment by fxd0h** [comment]
@coderabbitai full review

File: src/mflux/models/common_models/gpt_oss/gpt_oss.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+        if cache is None:
+            cache = [None] * len(self.layers)
+
+        full_mask = create_attention_mask(x, cache[self.ga_idx])
+        swa_mask = create_attention_mask(x, cache[self.swa_idx], window_size=self.window_size)
+
+        for layer, c, layer_type in zip(self.layers, cache, self.layer_types):
+            mask = full_mask if layer_type == "full_attention" else swa_mask
+            x = layer(x, mask, c)
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common_models/gpt_oss/gpt_oss.py]: Applied in 505a7d2, all three sites.

File: tests/test_lens_encoder.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            num_attention_heads=8,
+            num_key_value_heads=2,
+            intermediate_size=64,
+            num_local_experts=4,
+            num_experts_per_tok=2,
+            vocab_size=128,
+            sliding_window=4,
+            layer_types=["sliding_attention", "full_attention"] * (layers // 2),
+        )
+        return GptOssModel(args)
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply tests/test_lens_encoder.py:59]: Fixed in 505a7d2, the list is now built to the exact length.

File: src/mflux/models/lens/cli/lens_generate.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from mflux.utils.dimension_resolver import DimensionResolver
+from mflux.utils.exceptions import ModelConfigError, PromptFileReadError, StopImageGenerationException
+from mflux.utils.prompt_util import PromptUtil
+
+# Single source of truth for options this CLI accepts but cannot honour: the runtime
+# warning and the mflux-capabilities dump both read it.
+IGNORED_OPTIONS = {
+    "--guidance": "Lens Turbo is a 4-step distillation with CFG internalized; guidance is never applied.",
+    "--negative-prompt": "CFG is disabled on Lens Turbo, so the negative prompt is never encoded.",
+}
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/lens/cli/lens_generate.py:13]: Not wired here by design: the warning handler and mflux-capabilities live in #499, which reads these constants once it lands.

File: src/mflux/models/lens/model/text_encoder/lens_gpt_oss_encoder.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        stacked = mx.stack(captured, axis=2)  # [B, S, L, H]
+        return stacked[:, LENS_TXT_OFFSET:]
+
+    def _template_ids(self, prompt: str) -> list[int]:
+        # Harmony control markers inside the prompt would forge template blocks
+        # and desynchronize the fixed 97-token offset; strip them.
+        for marker in ("<|start|>", "<|end|>", "<|message|>", "<|channel|>", "<|return|>"):
+            prompt = prompt.replace(marker, " ")
+
+        encode = lambda text: self.tokenizer.encode(text, add_special_tokens=False).ids  # noqa: E731
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/lens/model/text_encoder/lens_gpt_oss_encoder.py:117]: Right, and the vocabulary check makes the case stronger than the list: against this checkpoint's tokenizer, `<|constrain|>`, `<|call|>` and `<|endoftext|>` are single ids (200003 / 200012 / 199999) with `add_special_tokens=False`, so all three were injectable. `<|refusal|>` is not in this vocab at all (it tokenizes as six ordinary pieces), which is the argument against a kept list: it differs per checkpoint. 505a7d2 strips the whole `<|...|>` shape instead, and the injection test covers the three real ones plus an invented marker.

File: src/mflux/models/lens/model/text_encoder/lens_gpt_oss_encoder.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        # Over-long prompt: truncating the RENDERED sequence would cut the fixed
+        # assistant tail and misalign the template. Truncate the prompt segment
+        # instead, keeping the frozen prefix and suffix intact.
+        empty = render_lens_chat("")
+        cut = empty.index("<|end|><|start|>assistant")
+        prefix_ids = encode(empty[:cut])  # system + developer + user header (the 97-token offset)
+        suffix_ids = encode(empty[cut:])  # user close + both assistant blocks
+        budget = LENS_MAX_TOKENS - len(prefix_ids) - len(suffix_ids)
+        prompt_ids = encode(prompt)[:budget]
+        return prefix_ids + prompt_ids + suffix_ids
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/lens/model/text_encoder/lens_gpt_oss_encoder.py:131]: Budget clamped in 505a7d2. The retokenization half does not reproduce: harmony markers are atomic special ids, so BPE merges cannot cross the `<|message|>` boundary. Splicing `prefix + encode(prompt) + suffix` gives token-identical output to encoding the whole rendered template, checked on ascii, accented and long-run prompts.

**Review by anthonywu — APPROVED** [review]
LGTM from me on interface contracts and module organiztion

would like another ML reviewer to give the 2nd ack before merge

**Comment by ianscrivener** [comment]
Microsoft Lens seems to have been WITHDRAWN nowhere to be found on HF?!
https://huggingface.co/microsoft/Lens
https://huggingface.co/microsoft/Lens-Base
https://huggingface.co/microsoft/Lens-Turbo
https://github.com/microsoft/Lens

**Comment by fxd0h** [comment]
We never pointed at microsoft/Lens. `lens-turbo` resolves to Comfy-Org/Lens, still up, around 20k downloads, converted from the microsoft repo per its own card.

So nothing breaks. The catch is that one mirror holds the weights now, and if that goes too the port has nowhere to point.

**Comment by fxd0h** [comment]
Found what Copilot was pointing at in #568, and it was real. Merging main into this branch silently ate two lines from the `lens-turbo` entry, `requires_sigma_shift=True` and its closing paren, so the dict entry ran into the next one and `model_config.py` stopped parsing. Git took it without a conflict because the incoming z-image controlnet entry ends in those same two lines.

Restored in f3eb61e. Module parses, fast suite green at 613.

Worth knowing for the other branches being updated today: this is the failure mode of adjacent entries in a big literal, no conflict markers and a syntax error instead.
